### DISCUSSION TITLE
How to append to environment variable in external graders?

### DISCUSSION THREAD

**Discussion by mfbutner** [PrairieLearn/PrairieLearn#6048] [Q&A]
I know you can use the environment option of externalGradingOptions in a question's info.json to set environment variables. Is it possible to use this feature to **append** to an environment variable? 

As for why I want to do this, I'm using the c/c++ autograder and I have a python file that I want to use across multiple questions.  As such, I plan on placing the file in serverFilesCourse but currently the c/c++ autograder doesn't add /grade/serverFilesCourse to `PYTHONPATH` but it does set `PYTHONPATH` to `/cgrader/`. This means I couldn't directly set  `PYTHONPATH` but need to append to it.

I am aware of the various workarounds to this problem
1. [Update path in my python script](https://stackoverflow.com/questions/5318441/is-it-possible-to-modify-pythonpath-at-runtime)
2. Set my entry point to be a bash script which first updates `PYTHONPATH` then runs my original program
3. Build my own Docker Image based on cgrader and update the path there
4. Set `PYTHONPATH` in info.json to  `/cgrader/;/grade/serverFilesCourse`. I don't like this because I don't own the cgrader autograder and it could update at any time with something that breaks my code. 

I feel like options 1 - 3 are just that annoying bit of extra work and don't document the dependency as well as setting the environment value in info.json but I can do any of them if I have to. I was just hoping that someone might know a way to do my preferred option.

**Comment by jonatanschroeder** [comment]
I wouldn't mind approving a PR that adds this directory to the C grader, it seems pretty sensible to add it there.

**Reply by mfbutner** [reply]
Well, let me get on that then. Shouldn't be too tough.

**Reply by mfbutner** [reply]
I think I've updated things correctly but am not sure what the best way to test my change is. Any suggestions?

 What I changed was `ENV PYTHONPATH=/cgrader/` to `ENV PYTHONPATH=/cgrader/:/grade/serverFilesCourse` in PrairieLearn/graders/python/Dockerfile.

**Reply by jonatanschroeder** [reply]
Here's what I usually do (assuming `~/pl` is where PrairieLearn is cloned):

1. Build the docker image, tagging it locally: `docker build ~/pl/graders/c -t prairielearn/grader-c`.
2. Create (or update) `~/pl/config.json` with the following:
```js
{
  # ... any other properties you may have
  "externalGradingPullImagesFromDockerHub": false
}
```
3. Run PrairieLearn with the git source (https://prairielearn.readthedocs.io/en/latest/installingLocal/).
4. Test the questions that have the features you need. The config option above will ensure you use the local version of the docker image, instead of pulling it from Docker Hub.

**Reply by mfbutner** [reply]
Thanks for the info. I was able to test my work and my changes worked. I also updated the base image of ubuntu to 22.04 and increased the version of python from 3.8 to 3.10 as I was using some 3.10 features in my programs. Are you ok with those additional changes?

**Reply by jonatanschroeder** [reply]
Those are already covered by #5995, which is being held to be merged between terms to avoid issues during final exams in summer terms.
